src/f_finale.c

```diff
@@ -79,6 +82,153 @@ static int midstage;                 // whether we're in "mid-stage"
+    frame.sound = JS_GetIntegerValue(js_frame, "sound");
+
+    array_push(*frames, frame);
+    (*framecount)++;
+}
+
+static cast_anim_t ParseEndFinale_CastAnims(json_t *js_castanim_entry, const char *lump)
+{
+    cast_anim_t out = {0};
+    out.name = D_GetStringForMnemonic(JS_GetStringValue(js_castanim_entry, "name"));
+    out.alertsound = JS_GetIntegerValue(js_castanim_entry, "alertsound");
+
+    json_t *js_alive_frame_list = JS_GetObject(js_castanim_entry, "aliveframes");
+    json_t *js_alive_frame = NULL;
+    JS_ArrayForEach(js_alive_frame, js_alive_frame_list)
+    {
+        ParseEndFinale_CastFrame(js_alive_frame, &out.aliveframes, &out.aliveframescount, lump);
+    }
+
+    json_t *js_death_frame = NULL;
+    json_t *js_death_frame_list = JS_GetObject(js_castanim_entry, "deathframes");
+    JS_ArrayForEach(js_death_frame, js_death_frame_list)
+    {
+        ParseEndFinale_CastFrame(js_death_frame, &out.deathframes, &out.deathframescount, lump);
+    }
+
+    return out;
+}
+
+static void ParseEndFinale_CastRollCall(json_t *js_castrollcall, end_finale_t *out, const char *lump)
+{
+    json_t *js_castanim_list = JS_GetObject(js_castrollcall, "castanims");
+
+    json_t *js_castanim_entry = NULL;
+    JS_ArrayForEach(js_castanim_entry, js_castanim_list)
+    {
+        cast_anim_t castanim_entry = ParseEndFinale_CastAnims(js_castanim_entry, lump);
+        out->cast_animscount++;
+        array_push(out->cast_anims, castanim_entry);
+    }
+}
+
+static void ParseEndFinale_Bunny(json_t *js_bunny, end_finale_t *out, const char *lump)
+{
+    out->bunny_overlay = JS_GetIntegerValue(js_bunny, "overlay");
+    out->bunny_overlaycount = JS_GetIntegerValue(js_bunny, "overlaycount");
+    out->bunny_overlaysound = JS_GetIntegerValue(js_bunny, "overlaysound");
+    out->bunny_overlayx = JS_GetIntegerValue(js_bunny, "overlayx");
+    out->bunny_overlayy = JS_GetIntegerValue(js_bunny, "overlayy");
+    const char *bunny_lump = JS_GetStringValue(js_bunny, "stitchimage");
+    if (bunny_lump == NULL || (W_CheckNumForName(bunny_lump) < 0))
+    {
+        I_Printf(VB_WARNING, "EndFinale: invalid bunny stitchimage field on lump '%s'", lump);
+    }
+    else
+    {
+        M_CopyLumpName(out->bunny_stitchimage, bunny_lump);
+    }
+}
+
+end_finale_t *F_ParseEndFinale(const char lump[9])
+{
+    // Does the JSON lump even exist?
+    json_t *json = JS_Open(lump, "finale", (version_t){1, 0, 0});
+    if (json == NULL)
+    {
+        I_Printf(VB_WARNING, "EndFinale: invalid lump '%s'", lump);
+        return NULL;
+    }
+
+    // Does the lump actually have any data?
+    json_t *data = JS_GetObject(json, "data");
+    if (JS_IsNull(data) || !JS_IsObject(data))
+    {
+        I_Printf(VB_WARNING, "EndFinale: data object undefined on lump '%s'", lump);
+        JS_Close(lump);
+        return NULL;
+    }
+
+    // Now, actually parse it
+    end_finale_t *out = Z_Malloc(sizeof(end_finale_t), PU_STATIC, NULL);
```

Comment:
The memory should be nullified (use `Z_Calloc`); otherwise, I get a crash with release optimisations.